servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/KeepAliveManager.java

```diff
@@ -0,0 +1,321 @@
+                            protected void channelIdle(final ChannelHandlerContext ctx, final IdleStateEvent evt) {
+                                onIdle.run();
+                            }
+                        }));
+    }
+
+    KeepAliveManager(final Channel channel, @Nullable final KeepAlivePolicy keepAlivePolicy,
+                     final Scheduler scheduler, final IdlenessDetector idlenessDetector) {
+        this.channel = channel;
+        this.scheduler = scheduler;
+        if (keepAlivePolicy != null) {
+            disallowKeepAliveWithoutActiveStreams = !keepAlivePolicy.withoutActiveStreams();
+            pingAckTimeoutMillis = keepAlivePolicy.ackTimeout().toMillis();
+            final GenericFutureListener<Future<? super Void>> goAwayListener = f -> {
+                if (f.isSuccess()) {
+                    LOGGER.debug("Closing channel={}, after keep-alive timeout.", this.channel);
+                    KeepAliveManager.this.close0();
+                }
+            };
+            pingWriteCompletionListener = future -> {
+                if (future.isSuccess() && keepAliveState == KEEP_ALIVE_ACK_PENDING) {
+                    // Schedule a task to verify ping ack within the pingAckTimeoutMillis
+                    keepAliveState = scheduler.afterMillis(() -> {
+                        if (keepAliveState != null) {
+                            keepAliveState = KEEP_ALIVE_ACK_TIMEDOUT;
+                            LOGGER.debug(
+                                    "channel={}, timeout {}ms waiting for keep-alive PING(ACK), writing go_away.",
+                                    this.channel, pingAckTimeoutMillis);
+                            channel.writeAndFlush(new DefaultHttp2GoAwayFrame(NO_ERROR))
+                                    .addListener(goAwayListener);
+                        }
+                    }, pingAckTimeoutMillis);
+                }
+            };
+            int idleInSeconds = (int) keepAlivePolicy.idleDuration().getSeconds();
+            idlenessDetector.configure(channel, idleInSeconds, this::channelIdle);
+        } else {
+            disallowKeepAliveWithoutActiveStreams = false;
+            pingAckTimeoutMillis = DEFAULT_ACK_TIMEOUT.toMillis();
+            pingWriteCompletionListener = null;
+        }
+    }
+
+    void pingReceived(final Http2PingFrame pingFrame) {
+        assert channel.eventLoop().inEventLoop();
+
+        if (pingFrame.ack()) {
+            long pingAckContent = pingFrame.content();
+            if (pingAckContent == GRACEFUL_CLOSE_PING_CONTENT) {
+                LOGGER.debug("channel={}, graceful close ping ack received.", channel);
+                cancelIfStateIsAFuture(gracefulCloseState);
+                gracefulCloseWriteSecondGoAway();
+            } else if (pingAckContent == KEEP_ALIVE_PING_CONTENT) {
+                cancelIfStateIsAFuture(keepAliveState);
+                keepAliveState = null;
+            }
```

Comment:
RFC says:

> Receivers of a PING frame that does not include an ACK flag MUST send a PING frame with the ACK flag set in response, with an identical payload.

Should we error ack frames with payload we do not expect?

Reply:
+1... I think this should be a connection error

Reply:
hmm I would be hesitant to assume that we are the only entity sending PINGs when we do not have the visibility of which all frames are sent on the wire.
If at all we do feel this is necessary, I think this should be part of an enforcement policy as opposed to the keep-alive policy as it exists in this PR.